### DISCUSSION TITLE
What does `CopyEmpty=` do?

### DISCUSSION THREAD

**Discussion by e-t-l** [sandboxie-plus/Sandboxie#4387] [Q&A Feedback]
The ini pref `CopyEmpty` was introduced at the same time as `DontCopy` and others, but I don't see any documentation on it besides the syntax to use, and unlike "Don't Copy," it isn't clear to me what "Copy Empty" is supposed to do. Can anyone explain it?

**Comment by offhub** [comment]
From ChatGPT:

The code defines an enumeration `ENUM_COPY_MODES` that specifies different modes of handling files during the file migration process. These modes determine whether and how files should be copied into the sandbox environment. Let's go through each mode in detail:

### 1. **FILE_DONT_COPY**
   - **Description**: This mode indicates that the file should **not** be copied into the sandbox under any circumstances.
   - **When It's Used**:
     - If the file path matches a pattern in the `DontCopy` list, the file will be excluded from migration.
     - If this mode is applied, the file is effectively blocked from being copied into the sandbox.
     - The `File_MigrateFile_GetMode` function will return this mode if there is a matching pattern (configured by the user) that specifies that certain files or paths should not be copied.
   - **Behavior**:
     - If the file is flagged with this mode, the system will deny writing to the sandbox location (e.g., an access denied error).
     - A log message will be generated, notifying the user that the file will not be copied, depending on the configuration setting `File_NotifyNoCopy`.
     - If the system is configured to block writes, the user is informed about the denied write operation (via message ID 2114).
     - If the system is set to allow reading but deny writing (e.g., when `File_MigrationDenyWrite` is `FALSE`), the system may open the file in a read-only mode and log the information (message ID 2115).

### 2. **FILE_COPY_CONTENT**
   - **Description**: This mode means that the file **should be copied along with its contents** into the sandbox.
   - **When It's Used**:
     - If the file path matches a pattern in the `CopyAlways` list, or if the file does not match any of the exclusion patterns but is under the size limit for copying, it will be copied with its contents into the sandbox.
     - The `File_MigrateFile_GetMode` function will return this mode for files that should always be copied (or files under a size limit).
   - **Behavior**:
     - The file is fully copied, including its data.
     - If the file is large (and larger than the configured `File_CopyLimitKb`), a prompt might be issued to the user asking whether they want to copy the file. This happens only if the `PromptForFileMigration` setting is enabled.
     - If the file is smaller than the size limit or the user agrees to copy the large file, the contents will be transferred to the sandbox.
     - A log message is issued indicating that the file is being copied (message ID 2113).

### 3. **FILE_COPY_EMPTY**
   - **Description**: This mode indicates that **only the file’s metadata (attributes) are copied**, but **not its contents**.
   - **When It's Used**:
     - If the file path matches a pattern in the `CopyEmpty` list, it will be copied as an empty file (with no content).
     - The `File_MigrateFile_GetMode` function will return this mode for files that should be copied with empty contents, typically for files where the actual content is either unnecessary or irrelevant for the sandboxed environment.
   - **Behavior**:
     - The file is created in the sandbox with the same metadata (such as attributes, timestamps) as the source file but without any actual data.
     - This can be useful for files that are meant to exist in the sandbox environment (such as placeholders or configuration files) but don’t need to contain any data.
     - A log message will be generated to inform the user that the file is being copied as an empty file (message ID 2113).

### Summary of Modes:

| Mode               | Description                                                      | When Used                                     | Behavior                                  |
|--------------------|------------------------------------------------------------------|-----------------------------------------------|-------------------------------------------|
| **FILE_DONT_COPY** | The file is **not copied** into the sandbox.                     | If the file matches a "Don't Copy" pattern.   | The file is blocked from being copied. Log a message about the block. |
| **FILE_COPY_CONTENT** | The file is **copied with its contents** into the sandbox.     | If the file matches a "Copy Always" pattern or is under the size limit. | The file and its contents are copied. User may be prompted for large files. |
| **FILE_COPY_EMPTY** | The file is copied, but **only the metadata** (attributes) are copied, **not the contents**. | If the file matches a "Copy Empty" pattern.   | The file is created in the sandbox with no content but retains attributes. |

### General Flow:
- When a file is being considered for migration into the sandbox, the system checks against the defined patterns in `FILE_MigrationOptions[NUM_COPY_MODES]` to determine which mode applies.
- The mode dictates whether the file's contents are copied, whether it's copied as an empty file, or if it should be excluded from migration altogether.
- Depending on the mode, the system may prompt the user for decisions (e.g., for large files) or log specific messages regarding the migration status.

These modes provide flexibility for managing which files are transferred into the sandbox, how they are treated, and under what conditions (like file size or specific patterns in the path).
